memiavl/export.go

```diff
@@ -2,113 +2,118 @@ package memiavl
-	if height > math.MaxUint32 {
-		return fmt.Errorf("height overflows uint32: %d", height)
-	}
-	version := uint32(height)
+type MultiTreeExporter struct {
+	// only one of them is non-nil
+	db    *DB
+	mtree *MultiTree
 
-	var mtree *MultiTree
-	if db.supportExportNonSnapshotVersion {
-		db, err := Load(db.dir, Options{
+	iTree    int
+	exporter *Exporter
+}
+
+func NewMultiTreeExporter(dir string, version uint32, supportExportNonSnapshotVersion bool) (exporter *MultiTreeExporter, err error) {
+	var (
+		db    *DB
+		mtree *MultiTree
+	)
+	if supportExportNonSnapshotVersion {
+		db, err = Load(dir, Options{
 			TargetVersion: version,
 			ZeroCopy:      true,
 			ReadOnly:      true,
 		})
 		if err != nil {
-			return errors.Wrapf(err, "invalid height: %d", height)
+			return nil, fmt.Errorf("invalid height: %d, %w", version, err)
 		}
-		defer func() {
-			if err := db.Close(); returnErr == nil {
-				returnErr = err
-			}
-		}()
-
-		mtree = &db.MultiTree
 	} else {
-		curVersion, err := currentVersion(db.dir)
+		curVersion, err := currentVersion(dir)
 		if err != nil {
-			return fmt.Errorf("failed to load current version: %w", err)
+			return nil, fmt.Errorf("failed to load current version: %w", err)
 		}
 		if int64(version) > curVersion {
-			return fmt.Errorf("snapshot is not created yet: height: %d", version)
+			return nil, fmt.Errorf("snapshot is not created yet: height: %d", version)
 		}
-		mtree, err = LoadMultiTree(filepath.Join(db.dir, snapshotName(int64(version))), true, 0)
+		mtree, err = LoadMultiTree(filepath.Join(dir, snapshotName(int64(version))), true, 0)
 		if err != nil {
-			return errors.Wrapf(err, "snapshot don't exists: height: %d", version)
+			return nil, fmt.Errorf("snapshot don't exists: height: %d, %w", version, err)
 		}
-		defer func() {
-			if err := mtree.Close(); returnErr == nil {
-				returnErr = err
-			}
-		}()
 	}
 
-	for _, tree := range mtree.trees {
-		if err := protoWriter.WriteMsg(&snapshottypes.SnapshotItem{
-			Item: &snapshottypes.SnapshotItem_Store{
-				Store: &snapshottypes.SnapshotStoreItem{
-					Name: tree.Name,
-				},
-			},
-		}); err != nil {
-			return err
-		}
+	return &MultiTreeExporter{
+		db:    db,
+		mtree: mtree,
+	}, nil
+}
 
-		exporter := tree.Tree.Export()
-		for {
-			node, err := exporter.Next()
-			if err == iavl.ErrorExportDone {
-				break
-			} else if err != nil {
-				return err
-			}
-			if err := protoWriter.WriteMsg(&snapshottypes.SnapshotItem{
-				Item: &snapshottypes.SnapshotItem_IAVL{
-					IAVL: &snapshottypes.SnapshotIAVLItem{
-						Key:     node.Key,
-						Value:   node.Value,
-						Height:  int32(node.Height),
-						Version: node.Version,
-					},
-				},
-			}); err != nil {
-				return err
-			}
+func (mte *MultiTreeExporter) trees() []NamedTree {
+	if mte.db != nil {
+		return mte.db.trees
+	}
+	return mte.mtree.trees
+}
+
+func (mte *MultiTreeExporter) Next() (interface{}, error) {
+	if mte.exporter != nil {
+		node, err := mte.exporter.Next()
```

Comment:
if the error happens, will the node be `non-nil` value? Otherwise, we could return directly.

Reply:
If there's error, the first value is always nil here.

Reply:
oh, you mean return directly, sounds good.

Reply:
actually, we need to handle the error ExportDone here, a little bit more complex.

Reply:
not sure why we need handle `mte.Next()` when `err == ErrorExportDone`

Reply:
when a single tree export finish, we need to move on to the next tree, it only stops when all trees exported.